### PR TITLE
`performInv_invs`

### PR DESCRIPTION
**Pull request by michaelmcinerney**
Signed-off-by: Michael McInerney <m.mcinerney@unsw.edu.au>

### CODE REVIEW

File: proof/refine/ARM/Syscall_R.thy [hunk 1]
```diff
@@ -858,6 +858,171 @@ lemma threadSet_not_curthread_ct_domain:
...
+      apply fastforce
+     apply (frule sc_ko_at_valid_objs_valid_sc')
+      apply fastforce
+     apply (clarsimp simp: valid_sched_context'_def scBits_simps obj_at_simps)
+    apply (wpsimp | wps)+
+  apply (clarsimp simp: ct_in_state'_def st_tcb_at'_def obj_at_simps runnable_eq_active')
+  done
+
+lemma tcbInReleaseQueue_update_st_tcb_at'[wp]:
+  "threadSet (tcbInReleaseQueue_update b) t \<lbrace>\<lambda>s. Q (st_tcb_at' P t' s)\<rbrace>"
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:966]: Probably unimportant if this hasn't been needed before, but maybe move this to TcbAcc_R?

Comment (corlewis) [reply proof/refine/ARM/Syscall_R.thy:966]: Some of the following more general lemmas could be moved earlier as well.

**Review by corlewis — APPROVED** [review]
Very nice!

File: proof/refine/ARM/Syscall_R.thy [hunk 2]
```diff
@@ -858,6 +858,171 @@ lemma threadSet_not_curthread_ct_domain:
...
+   apply (fastforce simp: obj_at_simps valid_tcb'_def tcb_cte_cases_def comp_def
+                          valid_sched_context'_def valid_sched_context_size'_def
+                          valid_release_queue'_def inQ_def)
+   done
+
+lemma setSchedContext_ct_in_state':
+  "\<lbrace>ct_in_state' P and sc_at' ptr\<rbrace> setSchedContext ptr sc \<lbrace>\<lambda>_. ct_in_state' P\<rbrace>"
+  apply (wpsimp wp: set_sc'.set_wp)
+  apply (fastforce simp: ct_in_state'_def st_tcb_at'_def obj_at_simps ps_clear_def)
+  done
```
Comment (mktnk3) [inline proof/refine/ARM/Syscall_R.thy:899]: I don’t think we need `sc_at’` in the precondition. We already have a lemma about `setSchedContext` and `pred_tcb_at’`. That with `hoare_lift_Pf` or something like that should work.

**Review by mktnk3 — APPROVED** [review]
Nice.

File: proof/refine/ARM/Syscall_R.thy (hunk 2 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Syscall_R.thy:899]: Done, and put in `SchedContext_R.thy`

**Pull request merged**